### PR TITLE
Fixes self-attention structured score

### PR DESCRIPTION
**Pull request by purujitgoyal**

### CODE REVIEW

File: structured_prediction_baselines/modules/structured_score/sequence_tagging/self_attention.py [hunk 1]
```diff
@@ -45,3 +66,31 @@ def forward(
 
         # reduction = "max" (Default)
         return attention_output.amax(dim=3).sum(2)
+
+    def _get_attention_mask(self, n_samples, mask: torch.Tensor):
+        device = 'cuda' if torch.cuda.is_available() else 'cpu'
+        batch_size, seq_length = mask.shape
+        attention_mask = torch.BoolTensor(batch_size * n_samples, seq_length, seq_length).fill_(False)
+        attention_mask = attention_mask.to(device=device)
+        masked_length = mask.sum(dim=-1)
```
Comment (leejayyoon) [inline structured_prediction_baselines/modules/structured_score/sequence_tagging/self_attention.py:75]: If the masked_length simply gives you the length of the contiguous 1s in the mask, e.g. 3 for the following mask [1110000000], then I think this code is correct.

File: structured_prediction_baselines/modules/structured_score/sequence_tagging/self_attention.py [hunk 2]
```diff
@@ -45,3 +66,31 @@ def forward(
...
+
+        for i in range(batch_size):
+            i_seq_len = masked_length[i]
+            for j in range(i_seq_len):
+                lower_idx, higher_idx = max(0, j - self.M), min(i_seq_len, j + self.M + 1)
+                attention_mask[i:i + n_samples, j, lower_idx:higher_idx] = True
```
Comment (leejayyoon) [inline structured_prediction_baselines/modules/structured_score/sequence_tagging/self_attention.py:81]: Don't we need to reshape this somewhere?  
(batch_size * n_samples, seq_length, seq_length) --> (batch_size, n_samples, seq_length, seq_length)

Comment (leejayyoon) [reply structured_prediction_baselines/modules/structured_score/sequence_tagging/self_attention.py:81]: Or do we simply change the input shape to be (reverse logic)
(batch_size * n_samples, seq_length, seq_length) from (batch_size, n_samples, seq_length, seq_length)? 
I couldn't find that piece of code yet.

Comment (leejayyoon) [reply structured_prediction_baselines/modules/structured_score/sequence_tagging/self_attention.py:81]: Ah, actually found it. You do it when you add `positional embeddings`. 
Separate comment up there (line 50).

File: structured_prediction_baselines/modules/structured_score/sequence_tagging/self_attention.py [hunk 3]
```diff
@@ -26,13 +45,15 @@ def forward(
     ) -> torch.Tensor:
         mask = buffer["mask"]
         batch_size, n_samples, seq_length, _ = y.shape
-        attention_mask = torch.BoolTensor(seq_length, seq_length).fill_(False)
-        for i in range(seq_length):
-            lower_idx, higher_idx = max(0, i - self.M), min(seq_length, i + self.M + 1)
-            attention_mask[i][lower_idx:higher_idx] = True
+        attention_mask = self._get_attention_mask(n_samples, mask)
+
+        if self._use_positional_encoding:
```
Comment (leejayyoon) [inline structured_prediction_baselines/modules/structured_score/sequence_tagging/self_attention.py:50]: Let's just reduce one line to
        attention_input = y.view(batch_size * n_samples, seq_length, -1) # change the shape to appropriate structure for self-attention function.
        if self._use_positional_encoding:
            attention_input = add_positional_features(attention_input)

**Review by leejayyoon — APPROVED** [review]

File: structured_prediction_baselines/modules/structured_score/sequence_tagging/self_attention.py (hunk 3 above)
Comment (leejayyoon) [reply structured_prediction_baselines/modules/structured_score/sequence_tagging/self_attention.py]: I directly committed actually reflecting the suggested changes above, does anybody know why I am not seeing the change in this PR?

**Review by leejayyoon — APPROVED** [review]

File: structured_prediction_baselines/modules/structured_score/sequence_tagging/self_attention.py (hunk 1 above)
Comment (zhiyangxu-umass) [reply structured_prediction_baselines/modules/structured_score/sequence_tagging/self_attention.py:76]: right, a line of comment to explain the shape and values of mask will be helpful.

**Review by zhiyangxu-umass — APPROVED** [review]

**Comment by leejayyoon** [comment]
Adding description to point out the initial problem related to the [commit c9dec0](https://github.com/dhruvdcoder/structured_prediction_baselines/pull/64/commits/c9dec0c63823ec31697e815a35001108ebb9296e)  to reduce future confusion on what the problem was in the first palce.

## The problem ##
The previous code did not take into account that batch of sequences is filled with "pad" tokens. 
The seq_length is the "shape" of the batch "with pad tokens". And their energy calculation became faulty because of this problem. 

For example, if we have 5 tokens, and a mask size of 3, we should calculate an energy of [1,2,3], [2,3,4], and [3,4,5]. But, if there is padding tokens and the whole sequence size is 7, it will end up calculating [1,2,3], [2,3,4], [3,4,5] and extra terms: [4,5,6], [5,6,7] which is faulty.
